### ISSUE TITLE
When the user presses Tab + Shift in modal dialog and the first element is focused the last focusable element should be focused 

### ISSUE THREAD

**Issue by bashibozuk** [vmware/clarity#2400]
```
[x] bug
[x] enhancement
```

### Expected behavior

When I open a modal then press Tab the first focusable element gets focused 
Then when I press Tab + Shift the last focusable element should get focused.
If I keep pressing Tab + Shift the previous focusable element should be focused.
Expected behavior is described here: https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html

### Actual behavior

When I open a modal then press Tab the first focusable element gets focused 
Then when I press Tab + Shift the modal itself gets focused.
If I keep pressing Tab + Shift nothing happens.
 
### Reproduction of behavior
Go to https://clarity-light-theme-v11-9rttaf.stackblitz.io/ and click the Open Modal button
A modal will show
Press Tab button then press Tab + Shift

  
  
  
  

### Environment details

* **Angular version:** 4.x.x

* **Clarity version:** 0.11.21

* **OS and version:** OSX 10.12.6 | Windows 10

* **Browser:** [Chrome 67.0.3396.99 | Firefox 60.0.2 | IE 11 ]

**Comment by gnomeontherun** [comment]
Our focus trap has a limitation currently. When the focus trap is setup on the modal, we listen for the `document:focusin` event (not specifically for tab or shift+tab key events), and when it is detected we check if the newly focused item is part of the focus trap element (ie the modal), we always reset it to focus on the modal itself. This makes sense as long as you are progressing down the page, as it resets the focus to the 'top' of the focus trap, but fails to wrap around if you are going from top to bottom. Pressing shift+tab will always trigger the focus trap to focus on the modal itself, ad infinitum.

jQuery has a `:tabbable` pseudo selector (https://api.jqueryui.com/tabbable-selector/) that selects elements that can be tabbed ([basically elements with a tabindex >=. 0](https://github.com/jquery/jquery-ui/blob/0b7246b6eeadfa9e2696e22f3230f6452f8129dc/ui/tabbable.js#L31)). So we could implement something similar to detect the last tabbable item in the focus trap, and compare the previously focused items to see if we are moving up or down the tree. We'd have to keep track of the last two focused elements to detect directionality (to determine if the user is using tab or shift+tab), but when we detect the focus has moved outside of the focus trap we still force focus back inside (either to the focus trap element or to its last tabbable element).

**Comment by youdz** [comment]
@gnomeontherun did not explain **why** we use `document.focusin` here, and I would understand if the first intuition of people reading this was "Just list to Tab or Shift+Tab keys!".

The problem is that not all devices use Tab to move focus around, so doing so would break the focus trap on these devices. Here is a quote from [MDN's documentation on accessibility](https://developer.mozilla.org/en-US/docs/Web/Accessibility/Keyboard-navigable_JavaScript_widgets):
>Don't assume that all focus changes will come via key and mouse events: assistive technologies such as screen readers can set the focus to any focusable element. Track focus using onfocus and onblur instead.

**Comment by vinayakpatil** [comment]
Hi everyone, can this be made available in release `v3.x`?
